crypto/x509/by_dir.c

```diff
@@ -327,10 +327,10 @@ static int get_cert_by_subject(X509_LOOKUP *xl, X509_LOOKUP_TYPE type,
         /*
          * we have added it to the cache so now pull it out again
          */
-        CRYPTO_THREAD_write_lock(ctx->lock);
+        CRYPTO_THREAD_write_lock(xl->store_ctx->lock);
         j = sk_X509_OBJECT_find(xl->store_ctx->objs, &stmp);
         tmp = sk_X509_OBJECT_value(xl->store_ctx->objs, j);
-        CRYPTO_THREAD_unlock(ctx->lock);
+        CRYPTO_THREAD_unlock(xl->store_ctx->lock);
```

Comment:
... and this to:

```
        X509_STORE_unlock(xl->store_ctx);
```

Reply:
Done. But as pointed out by @petrovr, there may be other places left in code where wrong lock is claimed.

Reply:
I did a quick check. This API is not used anywhere in code. Everywhere else it's just direct calls to CRYPTO_THREAD_lock(), CRYPTO_THREAD_unlock() on store ctx.

Reply:
That's weird, and quite unclear...